File: src/engine/adapters/sqlite/query.ts
import type { StatementSync } from 'node:sqlite';
import type { OpCtx } from '../adapter';
import { AdapterError, assertNotCancelled } from '../errors';
import type { SqliteHandle } from './client';
import { mapError } from './errors';

export type SqliteParam = string | number | bigint | null | Uint8Array;

// F9/D9: `prepare()` silently keeps only the first statement of a multi-statement string, with no
// error — `stmt.sourceSQL` is exactly the substring of `sql` that was actually compiled (verified:
// includes the original leading whitespace, excludes everything after the first statement's own
// trailing `;`), so whatever remains is the dropped tail. Whitespace, one trailing `;`, and SQL
// comments are the only things tolerated there; anything else means a second statement would have
// silently vanished, so this refuses instead — the console's own contract (§8.14) is one page per
// statement, and a dropped tail would mean a page that quietly describes half of what was asked.
function assertSingleStatement(sql: string, stmt: StatementSync): void {
  const remainder = sql
    .slice(stmt.sourceSQL.length)
    .replace(/--[^\n]*/g, '')
    .replace(/\/\*[\s\S]*?\*\//g, '')
    .replace(/[\s;]/g, '');
  if (remainder !== '') {
    throw new AdapterError(
      'E_QUERY',
      'multiple statements are not supported in a single statement',
    );
  }
}

// D3: a *user's own table data* statement gets setReadBigInts(true) — an INTEGER outside
// Number.MAX_SAFE_INTEGER otherwise throws ERR_OUT_OF_RANGE instead of just losing precision, and
// that would crash a whole page over one out-of-range value in one row. Catalog/pragma queries
// deliberately do NOT opt in: every integer field a pragma reports (cid, pk, seq, notnull,
// hidden, wr, unique, ...) is a small counter, and BigInt-ifying it would break the plain-number
// comparisons and JSON serialization catalog.ts's code relies on for no benefit — node:sqlite's
// own default (`readBigInts: false`) is already correct for those, so they simply don't ask.
export function prepareOne(
  h: SqliteHandle,
  sql: string,
  opts?: { readBigInts?: boolean },
): StatementSync {
  let stmt: StatementSync;
  try {
    stmt = h.db.prepare(sql);
  } catch (err) {
    throw mapError(err);
  }
  assertSingleStatement(sql, stmt);
  if (opts?.readBigInts) stmt.setReadBigInts(true);
  return stmt;
}

export interface QueryOptions {
  rowsAsArray?: boolean;
  logParams?: boolean;
  /** D3: set only by the read path's own row-fetching query — see prepareOne's own comment. */
  readBigInts?: boolean;
}

// The read/catalog path: SELECT-shaped, bounded (read.ts's own LIMIT pageSize+1), so `.all()`
// rather than the streaming `iterate()` console.ts uses for a user's own unbounded statement (D5).
export function runQuery<R = unknown>(
  h: SqliteHandle,
  sql: string,
  params: SqliteParam[],
  ctx: OpCtx,
  opts?: QueryOptions,
): R[] {
  ctx.setCommand(
    opts?.logParams && params.length > 0 ? `${sql} -- params: ${JSON.stringify(params)}` : sql,
  );
  assertNotCancelled(ctx);
  const stmt = prepareOne(h, sql, { readBigInts: opts?.readBigInts });
  if (opts?.rowsAsArray) stmt.setReturnArrays(true);
  try {
    return stmt.all(...params) as R[];
  } catch (err) {
    throw mapError(err);
  }
}

export interface CommandOptions {
  /** setCommand() was already called once for the whole batch (mutate.ts, P5 D9's precedent) —
   *  do not call it again per statement. */
  suppressCommand?: boolean;
}

// mutate.ts's counterpart to runQuery — INSERT/UPDATE/DELETE, never SELECT.
export function runCommand(
  h: SqliteHandle,
  sql: string,
  params: SqliteParam[],
  ctx: OpCtx,
  opts?: CommandOptions,
): { affectedRows: number } {
  if (!opts?.suppressCommand) ctx.setCommand(sql);
  assertNotCancelled(ctx);
  const stmt = prepareOne(h, sql);
  try {
    const result = stmt.run(...params);
    return { affectedRows: Number(result.changes) };
  } catch (err) {
    throw mapError(err);
  }
}

// mutate.ts's transaction control (BEGIN IMMEDIATE/COMMIT/ROLLBACK) — fixed adapter-internal
// literals, never user SQL, so they bypass prepareOne's single-statement guard (which would have
// nothing to guard against here) and go straight through exec().
export function execLiteral(h: SqliteHandle, sql: string): void {
  try {
    h.db.exec(sql);
  } catch (err) {
    throw mapError(err);
  }
}

File: shell/internal/adapters/sqlite/query.go
package sqlite

import (
	"context"
	"database/sql"
	"encoding/json"
	"regexp"

	"github.com/kirathecat/kira-studio/shell/internal/adapters"
)

// The three patterns assertSingleStatement strips from the remainder after the first top-level
// ';' — query.ts's own "whitespace, one trailing ';', and SQL comments are the only things
// tolerated there" (F9/D9), ported verbatim as regexes since the remainder, by construction, opens
// outside any string literal.
var (
	commentLine           = regexp.MustCompile(`--[^\n]*`)
	commentBlock          = regexp.MustCompile(`(?s)/\*.*?\*/`)
	whitespaceOrSemicolon = regexp.MustCompile(`[\s;]`)
)

// firstTopLevelSemicolon scans sqlText once, tracking comment/string-literal state, and returns
// the byte index of the first ';' that appears outside a comment or a quoted literal — or -1 if
// there is none. This is Go's answer to query.ts's own guard, which instead compared the full text
// against `StatementSync.sourceSQL` (the exact substring node:sqlite's own prepare() actually
// compiled) — database/sql exposes no such "how much did the driver consume" signal for any
// driver, modernc.org/sqlite included (verified against its own stmt.go: a multi-statement string
// is detected internally via SQLite's own pzTail, but that fact never crosses the driver.Stmt
// interface), so the boundary is found by parsing the SQL text directly instead.
func firstTopLevelSemicolon(s string) int {
	const (
		normal = iota
		lineComment
		blockComment
		single
		double
		backtick
	)
	state := normal
	for i := 0; i < len(s); i++ {
		c := s[i]
		switch state {
		case normal:
			switch {
			case c == '-' && i+1 < len(s) && s[i+1] == '-':
				state = lineComment
				i++
			case c == '/' && i+1 < len(s) && s[i+1] == '*':
				state = blockComment
				i++
			case c == '\'':
				state = single
			case c == '"':
				state = double
			case c == '`':
				state = backtick
			case c == ';':
				return i
			}
		case lineComment:
			if c == '\n' {
				state = normal
			}
		case blockComment:
			if c == '*' && i+1 < len(s) && s[i+1] == '/' {
				state = normal
				i++
			}
		case single:
			if c == '\'' {
				if i+1 < len(s) && s[i+1] == '\'' {
					i++
				} else {
					state = normal
				}
			}
		case double:
			if c == '"' {
				if i+1 < len(s) && s[i+1] == '"' {
					i++
				} else {
					state = normal
				}
			}
		case backtick:
			if c == '`' {
				if i+1 < len(s) && s[i+1] == '`' {
					i++
				} else {
					state = normal
				}
			}
		}
	}
	return -1
}

// assertSingleStatement is query.ts's own guard (F9/D9, re-derived per B9/SQ-1): whitespace, one
// trailing ';', and SQL comments after the first top-level ';' are the only things tolerated;
// anything else means a smuggled second statement — the console's own contract (§8.14) is one page
// per statement, and modernc.org/sqlite executes a multi-statement string in full (SQ-1) rather
// than silently dropping the tail the way node:sqlite's prepare() did, so the enforcement has to
// happen here, before the driver ever sees the text.
func assertSingleStatement(sqlText string) error {
	idx := firstTopLevelSemicolon(sqlText)
	if idx < 0 {
		return nil
	}
	remainder := commentLine.ReplaceAllString(sqlText[idx+1:], "")
	remainder = commentBlock.ReplaceAllString(remainder, "")
	remainder = whitespaceOrSemicolon.ReplaceAllString(remainder, "")
	if remainder != "" {
		return adapters.New(adapters.CodeQuery, "multiple statements are not supported in a single statement", nil)
	}
	return nil
}

func setCommand(op *adapters.OpCtx, sqlText string, params []any, logParams bool) {
	if logParams && len(params) > 0 {
		if b, err := json.Marshal(params); err == nil {
			op.SetCommand(sqlText + " -- params: " + string(b))
			return
		}
	}
	op.SetCommand(sqlText)
}

// runRows is query.ts's runQuery generalized with a per-row scan callback, the same shape
// mysqlfamily/catalog.go's queryExec uses — catalog.go's typed pragma reads and read.go's own
// positional row reads both go through this. ctx is always the adapter-owned driverCtx
// (adapter.go's runOnConn), never the op's own context — B8's whole point.
func runRows(ctx context.Context, conn *sql.Conn, sqlText string, params []any, op *adapters.OpCtx, logParams bool, scan func(*sql.Rows) error) error {
	setCommand(op, sqlText, params, logParams)
	if err := assertSingleStatement(sqlText); err != nil {
		return err
	}
	if err := adapters.CheckNotStarted(ctx); err != nil {
		return err
	}

	rows, err := conn.QueryContext(ctx, sqlText, params...)
	if err != nil {
		return mapError(err)
	}
	defer rows.Close()
	for rows.Next() {
		if err := scan(rows); err != nil {
			return mapError(err)
		}
	}
	if err := rows.Err(); err != nil {
		return mapError(err)
	}
	return nil
}

// runArrayQuery is read.ts's runQuery({rowsAsArray: true}) — every row scanned positionally into
// []any, the driver's own native Go value for each cell (nil/int64/float64/string/[]byte, plus the
// one coercion the driver performs itself for a DATE/DATETIME/TIMESTAMP-declared column — see
// read.go's selectExpr and toCellText). D3's setReadBigInts has no Go analogue to opt into: an
// int64 is already exact, so there is no equivalent to node:sqlite's ERR_OUT_OF_RANGE to avoid.
func runArrayQuery(ctx context.Context, conn *sql.Conn, sqlText string, params []any, op *adapters.OpCtx, logParams bool) ([][]any, error) {
	var out [][]any
	err := runRows(ctx, conn, sqlText, params, op, logParams, func(rows *sql.Rows) error {
		cols, err := rows.Columns()
		if err != nil {
			return err
		}
		vals := make([]any, len(cols))
		dest := make([]any, len(cols))
		for i := range vals {
			dest[i] = &vals[i]
		}
		if err := rows.Scan(dest...); err != nil {
			return err
		}
		out = append(out, vals)
		return nil
	})
	return out, err
}

// runCommand is mutate.ts's runQuery-for-writes counterpart — INSERT/UPDATE/DELETE, never SELECT.
// suppressCommand mirrors mutate.ts's own "setCommand() was already called once for the whole
// batch" rule (P5 D9's precedent).
func runCommand(ctx context.Context, conn *sql.Conn, sqlText string, params []any, op *adapters.OpCtx, suppressCommand bool) (int64, error) {
	if !suppressCommand {
		op.SetCommand(sqlText)
	}
	if err := assertSingleStatement(sqlText); err != nil {
		return 0, err
	}
	if err := adapters.CheckNotStarted(ctx); err != nil {
		return 0, err
	}

	result, err := conn.ExecContext(ctx, sqlText, params...)
	if err != nil {
		return 0, mapError(err)
	}
	n, err := result.RowsAffected()
	if err != nil {
		return 0, mapError(err)
	}
	return n, nil
}

// execLiteral is mutate.ts's execLiteral — BEGIN IMMEDIATE/COMMIT/ROLLBACK, fixed adapter-internal
// literals never seen by a user, so they bypass assertSingleStatement (nothing to guard against
// here) and go straight through ExecContext.
func execLiteral(ctx context.Context, conn *sql.Conn, sqlText string) error {
	if _, err := conn.ExecContext(ctx, sqlText); err != nil {
		return mapError(err)
	}
	return nil
}
